File: WLogMerger/GUID.cpp
#include "GUID.h"
#include "ParameterException.h"
#include <sstream>

extern const std::string GUID::NULL_STRING("0000000000000000");
extern const char* GUID::TYPE_NAMES[8] = {
  "Player", "Creature", "Pet", "GameObject", "Vehicle", "Vignette", "Item", "Corpse"
};

GUID::GUID(std::string guid)
{
  switch(guid[0])
  {
    case 'P': type = guid[1] == 'l' ? TYPE::PLAYER : TYPE::PET; break;
    case 'C': type = guid[1] == 'r' ? TYPE::CREATURE : TYPE::CORPSE; break;
    case 'G': type = TYPE::GAME_OBJECT; break;
    case 'V': type = guid[1] == 'e' ? TYPE::VEHICLE : TYPE::VIGNETTE; break;
    case 'I': type = TYPE::ITEM; break;
    case '0': if(guid == NULL_STRING) { type = TYPE::NIL; break; } // else -> fall through
    default: throw ParameterException("Invalid GUID", guid);
  }

  size_t index;
  if(type == TYPE::PLAYER)
  {
    serverID = std::stoi(guid.substr(guid.find('-') + 1), &index, 16);
    spawnUID = std::stoi(guid.substr(index + 1), &index, 16);
    if(index != guid.length())
      throw ParameterException("Invalid GUID", guid);
  }
  else if(type == TYPE::ITEM)
  {
    serverID = std::stoi(guid.substr(guid.find('-') + 1), &index, 16);
    if(guid[index] != '-' || guid[index + 1] != '0' || guid[index + 2] != '-')
      throw ParameterException("Invalid GUID", guid);
    spawnUID = std::stoi(guid.substr(index + 3), &index, 16);
    if(index != guid.length())
      throw ParameterException("Invalid GUID", guid);
  }
  else if(type != TYPE::NIL)
  {
    index = guid.find('-');
    if(guid[index + 1] != '0' || guid[index + 2] != '-')
      throw ParameterException("Invalid GUID", guid);
    serverID   = std::stoi(guid.substr(index + 3), &index, 16);
    instanceID = std::stoi(guid.substr(index + 1), &index, 16);
    zoneUID    = std::stoi(guid.substr(index + 1), &index, 16);
    ID         = std::stoi(guid.substr(index + 1), &index, 16);
    spawnUID   = std::stoi(guid.substr(index + 1), &index, 16);
    if(index != guid.length())
      throw ParameterException("Invalid GUID", guid);
  }
}

std::string GUID::getString() const
{
  if(type == TYPE::NIL)
    return NULL_STRING;
    
  std::stringstream ss;
  ss.fill('0');
  ss.setf(std::ios_base::hex , std::ios_base::basefield);
  ss << TYPE_NAMES[(int) type];

  if(type == TYPE::PLAYER)
  {
    ss << '-' << serverID << '-';
    ss.width(8);
    ss << spawnUID;
  }
  else if(type == TYPE::ITEM)
  {
    ss << '-' << serverID << "-0-";
    ss.width(16);
    ss << spawnUID;
  }
  else
  {
    ss << "-0-" << serverID << '-' << instanceID << '-' << zoneUID << '-' << ID << '-';
    ss.width(10);
    ss << spawnUID;
  }

  return ss.str();
}

bool GUID::operator==(const GUID& other) const
{
  if(type != other.type) return false;
  if(type == TYPE::NIL) return true;
  if(type == TYPE::PLAYER || type == TYPE::ITEM) return spawnUID == other.spawnUID;
  return spawnUID == other.spawnUID && 
         instanceID == other.instanceID && 
         zoneUID == other.zoneUID && 
         ID == other.ID;
}

File: wlog/GUID.py
from enum import Enum

# For players: Player-[server ID]-[player UID] (Example: "Player-970-0002FD64")
# For creatures, pets, objects, and vehicles: [Unit type]-0-[server ID]-[instance ID]-[zone UID]-[ID]-[spawn UID] (Example: "Creature-0-970-0-11-31146-000136DF91")
#   Unit Type Names: "Creature", "Pet", "GameObject", "Vehicle", and "Vignette"
# For items: Item-[server ID]-0-[spawn UID] (Example: "Item-970-0-400000076620BFF4") (Please note that this tells you nothing useful about the item, like the ID)
class GUID:
    def __init__(self, guid: str):
        r = guid.split('-')

        # Check GUID Type
        if len(r) == 1 and int(r[0]) == 0:
            self.type = GUID.TYPE.NULL
        elif r[0] == 'Player':
            self.type = GUID.TYPE.PLAYER
        elif r[0] == 'Creature':
            self.type = GUID.TYPE.CREATURE
        elif r[0] == 'Pet':
            self.type = GUID.TYPE.PET
        elif r[0] == 'GameObject':
            self.type = GUID.TYPE.GAME_OBJECT
        elif r[0] == 'Vehicle':
            self.type = GUID.TYPE.VEHICLE
        elif r[0] == 'Vignette':
            self.type = GUID.TYPE.VIGNETTE
        elif r[0] == 'Item':
            self.type = GUID.TYPE.ITEM
        elif r[0] == 'Corpse':
            # Corpse-0-4445-409-24916-0-0000067792
            self.type = GUID.TYPE.CORPSE
        else:
            raise ValueError('Invalid GUID Type: ' + guid)
        
        # Read GUID IDs
        if self.type == GUID.TYPE.PLAYER:
            assert(len(r) == 3)
            assert(len(r[2]) == 8)
            self.serverID   = int(r[1], 16)
            self.playerUID  = int(r[2], 16)
        elif self.type == GUID.TYPE.ITEM:
            assert(len(r) == 4)
            assert(r[2] == '0')
            assert(len(r[3]) == 16)
            self.serverID   = int(r[1], 16)
            self.spawnUID   = int(r[3], 16)
        elif self.type != GUID.TYPE.NULL:
            assert(len(r) == 7)
            assert(r[1] == '0')
            assert(len(r[6]) == 10)
            self.serverID   = int(r[2], 16)
            self.instanceID = int(r[3], 16)
            self.zoneUID    = int(r[4], 16)
            self.ID         = int(r[5], 16)
            self.spawnUID   = int(r[6], 16)
        else:
            if self.type != GUID.TYPE.NULL:
                print(self.type)
                assert(False and 'GUID has no or invalid Type?')

    def __str__(self):
        if self.type == GUID.TYPE.NULL:
            return '0000000000000000'
        if self.type == GUID.TYPE.PLAYER:
            return 'Player-%X-%08X' % (self.serverID, self.playerUID)
        if self.type == GUID.TYPE.CREATURE:
            return 'Creature-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.PET:
            return 'Pet-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.GAME_OBJECT:
            return 'GameObject-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.VEHICLE:
            return 'Vehicle-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.VIGNETTE:
            return 'Vignette-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.CORPSE:
            return 'Corpse-0-%X-%X-%X-%X-%010X' % (self.serverID, self.instanceID, self.zoneUID, self.ID, self.spawnUID)
        if self.type == GUID.TYPE.ITEM:
            return 'Item-%X-0-%016X' % (self.serverID, self.spawnUID)
        print(self.type)
        assert(False and 'GUID has no or invalid Type?')

    def __eq__(self, other):
        if not isinstance(other, GUID):
            return False
        if self.type != other.type:
            return False
        if self.type == GUID.TYPE.NULL:
            return True
        if self.serverID != other.serverID:
            return False
        if self.type == GUID.TYPE.PLAYER:
            return self.playerUID == other.playerUID
        if self.type == GUID.TYPE.ITEM:
            return self.spawnUID == other.spawnUID
        return self.instanceID == other.instanceID and \
               self.zoneUID == other.zoneUID and \
               self.ID == other.ID and \
               self.spawnUID == other.spawnUID
    
    def __ne__(self, other):
        return not self.__eq__(other)

    class TYPE(Enum):
        NULL = 0
        PLAYER = 1
        CREATURE = 2
        PET = 3
        GAME_OBJECT = 4
        VEHICLE = 5
        VIGNETTE = 6
        ITEM = 7
        CORPSE = 8